Add new ASA profile (by Akko)

lvisintini opened the pull request:
## Background

I notices this profile was not available in the codebase.
According to [this guide](https://epomaker.com/blogs/guides/keycaps-profiles-overview-what-are-they-and-how-to-choose), "The ASA profile mixes the height of the OEM profile and maintains the spherical shape of keycaps. It's a customized profile provided by the AKKO."

Some additional information on the internet revealed a few more details on the profile:

![9a05f0c6-fab2-4d71-858f-4e6ad1d50be8 __CR0,0,970,300_PT0_SX970_V1___](https://user-images.githubusercontent.com/3267001/131343849-095414f5-44d7-4cc1-8cf0-1ea398d9a1e5.jpg)
![bc4647c9-25d6-4642-ac4c-785fb0320f38 __CR0,0,970,600_PT0_SX970_V1___](https://user-images.githubusercontent.com/3267001/131343854-988ec79d-96c8-4f97-b722-d95f5d14fde5.jpg)
I have my own set of ASA keys and, using callipers, I found the height of the top is 12mm.

Knowing all that I was able to calculate the tilts for the tops for each row.

I measured the bottom of 10 keys for each row with calipers (Width and Height) and then averaged the results.

Base Height | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Average
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
R1 | 18.07 | 18.06 | 18.07 | 18.07 | 18.07 | 18.06 | 18.07 | 18.07 | 18.07 | 18.07 | 18.068
R2 | 18.06 | 18.05 | 18.06 | 18.06 | 18.05 | 18.05 | 18.04 | 18.07 | 18.05 | 18.06 | 18.055
R3 | 18.06 | 18.07 | 18.07 | 18.09 | 18.07 | 18.06 | 18.07 | 18.07 | 18.08 | 18.07 | 18.071
R4 | 18.05 | 18.08 | 18.06 | 18.05 | 18.06 | 18.07 | 18.07 | 18.04 | 18.06 | 18.05 | 18.059
  |   |   |   |   |   |   |   |   |   | Overall average | 18.063


Base Width | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Average
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
R1 | 17.95 | 17.96 | 17.94 | 17.94 | 17.95 | 17.96 | 17.96 | 17.96 | 17.96 | 17.96 | 17.9540
R2 | 18.19 | 18.17 | 18.19 | 18.09 | 18.17 | 18.09 | 18.19 | 18.19 | 18.18 | 18.19 | 18.1650
R3 | 18.05 | 18.07 | 18.05 | 18.05 | 18.06 | 18.04 | 18.05 | 18.06 | 18.04 | 18.05 | 18.0520
R4 | 18.02 | 18.04 | 18.00 | 18.01 | 18.02 | 18.02 | 18.02 | 18.01 | 18.00 | 18.01 | 18.0150
  |   |   |   |   |   |   |   |   |   | Overall average | 18.0465


I think that the heights seem to be mostly consistent across the board.
This I would say it is reasonable to use the overall average value of 18.06 (rounded)

For the widths, the values for each row as not as consistent, so I would use the average values for each row and use the overall average value of 18.05 (again rounded) as a default (for row 5 and beyond)

  | Base Height | Base Width
-- | -- | --
R1 | 18.06 | 17.95
R2 | 18.06 | 18.17
R3 | 18.06 | 18.05
R4 | 18.06 | 18.02
Default | 18.06 | 18.05


The assumption is that everything else about the shape of the keys is the same as the SA profile.

 The below code produces the below output:

```
include <KeyV2/includes.scad>

legends = [
    "R1",
    "R2",
    "R3",
    "R4",
];

for(row=[1:4]) {
  translate_u(0,row)
  asa_row(row=row)
  legend(legends[row-1], [0,0, 6])
  key();
}
```

![preview](https://user-images.githubusercontent.com/3267001/140612627-043f619e-3bfe-4041-aa00-6f3a8386bb1e.png)


## Changes
- Create src/key_profiles/asa.scad to handle config for ASA profile
- Add new profile to src/key_profiles.scad

lvisintini commented:
Dish sizes and other parts of the config for this profile may be a bit off.
This is because:
- I do not know what dish size is or how to measure it (Please let me know how you measured yours)
- I only made the modifications required to match the information I found online. The gaps in data were filled using the configuration for OEM (dish dimensions among them)

Could you confirm that I am understanding your comment appropriately?
> Dishes a little bit bigger than original 13.26mm vs 12.20mm on R4 for ex. and dishes on original are dipper.

You are saying that the dish size of your tests print is a bit bigger than the ones you have in your "akko retro set"
For example, your R4 print had a 13.26mm dish size and the one in your akko retro set is 12.20mm instead.

Is that the correct interpretation of your comment?

I have the [this one](https://ltcofficial.com/collections/keyboard/products/ltc-lavacaps-pbt-doubleshot-146-keycaps-set-translucent-dsa-profile-for-mechanical-keyboard-ansi-us-layout-61-68-84-87-tkl-104-108-keys-with-keycap-puller-storage-tray-only-keycaps-white)

They are not a "retro" set, so I assume they are more standard... maybe
I did some test in my Prusa Mk3. Ignoring the layering you would find on any FDM print, the keys were very similar.
It could be your retro set is similar to this, but ultimately a variation of it.

Ultimately... my set is an ASA profile made by LTC and yours is an ASA profile made by Akko. there could be differences.
I do not have an Akko set to compare

limitium commented:
`Is that the correct interpretation of your comment?` - right
This is mine https://aliexpress.ru/item/1005002220211200.html?sku_id=12000023278543260&spm=a2g2w.productlist.0.0.bb716006rql0pR.

I'll make some shots with direct compare of the caps

lvisintini commented:
@limitium I measured 10 keys for each row with calipers (Width and Height) and then averaged the results.

Height | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Average
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
R1 | 18.07 | 18.06 | 18.07 | 18.07 | 18.07 | 18.06 | 18.07 | 18.07 | 18.07 | 18.07 | 18.068
R2 | 18.06 | 18.05 | 18.06 | 18.06 | 18.05 | 18.05 | 18.04 | 18.07 | 18.05 | 18.06 | 18.055
R3 | 18.06 | 18.07 | 18.07 | 18.09 | 18.07 | 18.06 | 18.07 | 18.07 | 18.08 | 18.07 | 18.071
R4 | 18.05 | 18.08 | 18.06 | 18.05 | 18.06 | 18.07 | 18.07 | 18.04 | 18.06 | 18.05 | 18.059
  |   |   |   |   |   |   |   |   |   | Overall average | 18.063


Width | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | Average
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
R1 | 17.95 | 17.96 | 17.94 | 17.94 | 17.95 | 17.96 | 17.96 | 17.96 | 17.96 | 17.96 | 17.9540
R2 | 18.19 | 18.17 | 18.19 | 18.09 | 18.17 | 18.09 | 18.19 | 18.19 | 18.18 | 18.19 | 18.1650
R3 | 18.05 | 18.07 | 18.05 | 18.05 | 18.06 | 18.04 | 18.05 | 18.06 | 18.04 | 18.05 | 18.0520
R4 | 18.02 | 18.04 | 18.00 | 18.01 | 18.02 | 18.02 | 18.02 | 18.01 | 18.00 | 18.01 | 18.0150
  |   |   |   |   |   |   |   |   |   | Overall average | 18.0465


I think that the heights seem to be mostly consistent across the board.
This I would say it is reasonable to use the overall average value of 18.06 (rounded)

For the widths, the values for each row as not as consistent, so I would use the average values for each row and use the overall average value of 18.05 (again rounded) as a default (for row 5 and beyond)

  | Height | Width
-- | -- | --
R1 | 18.06 | 17.95
R2 | 18.06 | 18.17
R3 | 18.06 | 18.05
R4 | 18.06 | 18.02
Default | 18.06 | 18.05

With height and width well defined, my hope is that the dishes should better match the stock ones, as I think that those are done in proportion to the overall dimensions.

I will add this values in and then do a few tests prints. then I will let you know how it goes

lvisintini commented:
@limitium, I have updated the height and width according to my latest comment.

I made changes to the dish size and dish depth as well, after some careful measurements with callipers revealed that I was a bit off on those.

I have also cleaned up the code a bit as there were some redundancies.

**I will this PR description with all this chances.** It should make it easy for anyone else reviewing.

My tests prints look okay to me now.

limitium commented:
Hey, I've measured Akko version of ASA profile with following averages
| |Width|Height middle|Height corner|
|-------|-------|-------|-------|
|R1|17.96|11.55|11.90|
|R2|17.98|9.99|10.40|
|R3|18.04|9.31|9.61|
|R4|17.85|10.40|10.60|

I don't get how did you have same height for all rows, because they definitely have a different heights, look into official specs
![](https://m.media-amazon.com/images/S/aplus-media/sc/9a05f0c6-fab2-4d71-858f-4e6ad1d50be8.__CR0,0,970,300_PT0_SX970_V1___.jpg)

What I've measured:

- Height middle - the maximum height in the middle vertical line of the keycap
- Height corner - maximum height of the corner of the dish

My printer is busy for next couple of days, I'll check PRs asap. Also I've asked friend of mine to print these keycaps on resin printer with the finest resolution.

lvisintini commented:
@limitium Sorry. perhaps I used the wrong words to describe my measurements.

When I wrote "**Height**", I meant to say the length of the base of the keycap along the Y axis (referred to as `bottom_key_height`)

When I wrote "**Width**", I meant to say the length of the base of the keycap along the X axis (referred to as `bottom_key_width `)

The dimensions of the keycap along the Z axis are as described on the official specs.

limitium commented:
I've printed new version and there some issues, probably they can be printed related however this result I have:
* There is a difference in shape especially R1 and R4
* Dishes are larger than original
* Dishes sphere curvature is less than original

![photo_2021-11-10_19-52-18](https://user-images.githubusercontent.com/850285/141157005-428628a4-2abb-44c4-8acd-541c9f2d3df3.jpg)
![photo_2021-11-10_19-52-16](https://user-images.githubusercontent.com/850285/141157010-3e9fa8b8-c2dc-4324-ac47-d21518e917ad.jpg)
![photo_2021-11-10_19-52-12](https://user-images.githubusercontent.com/850285/141157013-9342fd85-6ad8-451d-a2dd-c4749a83b60c.jpg)
![photo_2021-11-10_19-52-10](https://user-images.githubusercontent.com/850285/141157017-7a04b64f-1f6d-46c3-b60a-41f2cd27ca56.jpg)
![photo_2021-11-10_19-52-08](https://user-images.githubusercontent.com/850285/141157019-72d9e6a3-ac36-4e6f-8602-3eb3c8d87538.jpg)
![photo_2021-11-10_19-52-06](https://user-images.githubusercontent.com/850285/141157020-2cde2825-3803-4ac5-9913-a7d49168a83f.jpg)
![photo_2021-11-10_19-52-03](https://user-images.githubusercontent.com/850285/141157026-bbe9bb87-ef71-429d-aed8-89421450d1a2.jpg)

Here you can see that dishes are avg 1mm wider than original
![photo_2021-11-10_19-52-02](https://user-images.githubusercontent.com/850285/141157027-fd39920e-813d-49b6-a6a7-7f1e7cd9b227.jpg)
![photo_2021-11-10_19-52-00](https://user-images.githubusercontent.com/850285/141157029-a691fb42-cc30-4edd-8f44-c30adf122af7.jpg)

Here you can see that R4 and R1 have another shape
![photo_2021-11-10_19-51-57](https://user-images.githubusercontent.com/850285/141157030-0e40e337-0c92-4eaa-8e6b-c2f06fabae99.jpg)
![photo_2021-11-10_19-51-55](https://user-images.githubusercontent.com/850285/141157031-91153bfb-4fab-4a16-887a-a9a8cb7b5385.jpg)
![photo_2021-11-10_19-51-51](https://user-images.githubusercontent.com/850285/141157036-795a740a-c3b6-4e0a-a956-ff4ef9dddb01.jpg)
![photo_2021-11-10_19-51-42](https://user-images.githubusercontent.com/850285/141157037-0d3bcf0b-5871-4c79-924f-148081686aba.jpg)

rsheldiii commented:
This library doesn't purport to be a perfect 1:1 replacement for keycaps, so while it'd be great to have these be perfectly accurate, I'm happy to merge this now. I'll add an issue card to get some ASA keycaps and see if I can't iterate on a more accurate shape
